### PR TITLE
[BEAM-7739] Implement ReadModifyWriteState in Python SDK

### PR DESCRIPTION
**Pull request by rakeshcusat**
ValueState is missing from python sdk though Java sdk has it. In order to have the feature parity we should have this implemented for Python sdk.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
Thanks for looking at this. IIRC, the consensus last time this came up was to call it ReadModifyWriteState.

**Comment by rakeshcusat** [comment]
@robertwb Can you give me more context to that discussion or email chain that I can follow. 
I was also talking to @pabloem offline and he mentioned that ValueState is controversial and java sdk doesn't allow it specifically when you are merging windows.

**Comment by rakeshcusat** [comment]
@robertwb never mind, I found the discussion [here](https://lists.apache.org/thread.html/ccc0d548e440b63897b6784cd7896c266498df64c9c63ce6c52ae098@%3Cdev.beam.apache.org%3E)

**Comment by robertwb** [comment]
To follow up on this, it's not yet implemented but that would be a welcome contribution (likely just some renaming in this PR).

**Comment by rakeshcusat** [comment]
@robertwb  can you confirm that we also need to introduce ReadModifyWriteState in the proto definition?
https://github.com/apache/beam/blob/d510c2b072d6c56d335b893e93c7767cb40ae828/model/pipeline/src/main/proto/beam_runner_api.proto#L414-L424

**Comment by robertwb** [comment]
> @robertwb can you confirm that we also need to introduce ReadModifyWriteState in the proto definition?

Or simply rename ValueStateSpec in the proto. (These protos are still in a state they can be modified.)

**Comment by rakeshcusat** [comment]
> ValueStateSpec

I just want to make sure I understand it correctly, In case I rename this, I also need to rename in Java SDK as well because it is used there as well, right? 
I didn't see any reference in Go SDK other then the auto-generated code so I am assuming it is not yet implemented in Go SDK

**Comment by rakeshcusat** [comment]
> > @robertwb can you confirm that we also need to introduce ReadModifyWriteState in the proto definition?
> 
> Or simply rename ValueStateSpec in the proto. (These protos are still in a state they can be modified.)

I just want to make sure I understand it correctly, In case I rename this, I also need to rename in Java SDK as well because it is used there as well, right? 
I didn't see any reference in Go SDK other then the auto-generated code so I am assuming it is not yet implemented in Go SDK

I will rename the spec to be consistent and avoid any confusion in future.

**Comment by pabloem** [comment]
@rakeshcusat I think the name for Java would still be the same (changing the name would be a breaking change).

As for the proto, I think you can change it, because it's not externally exposed to users of Beam - it's only meant to be used by Beam itself (in this repository).

**Comment by robertwb** [comment]
Ping?

**Comment by rakeshcusat** [comment]
> Ping?

@robertwb I was busy but I will try to finish it this week.

**Comment by rakeshcusat** [comment]
@robertwb / @pabloem Here is the current status:
1. Updated proto `beam_runner_api.proto`
2. Python SDK side is ready.
3. Made most of the changes in Java SDK and I need feedback to not break the existing ValueState but deprecate it with minimal changes. I would really appreciate some pointers and ideas.
4. I have not looked into Go SDK and not sure whether it is implemented or not but would appreciate some pointers and ideas.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryStateInternals.java [hunk 1]
```diff
@@ -166,9 +172,62 @@ public WatermarkHoldState bindWatermark(
...
+    @Override
+    public boolean isCleared() {
+      return isCleared;
+    }
+  }
+
+  /** An {@link InMemoryState} implementation of {@link ReadModifyWriteState}. */
   public static final class InMemoryValue<T>
-      implements ValueState<T>, InMemoryState<InMemoryValue<T>> {
+          implements ValueState<T>, InMemoryState<InMemoryValue<T>> {
```
[the remark this answers is not in the document]
Comment (rakeshcusat) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryStateInternals.java]: This is a good idea, I need to look more closely and consolidate duplicate code.

**Comment by rakeshcusat** [comment]
R: @kennknowles

**Comment by robertwb** [comment]
Sound good to me. It'd be less work to swap steps (1) and (2).

**Comment by rakeshcusat** [comment]
> Sound good to me. It'd be less work to swap steps (1) and (2).

Sounds good. I will do as the suggested way.

**Comment by robertwb** [comment]
Any update on this? I see https://github.com/apache/beam/pull/9687, were there other PRs?

**Pull request closed without merging**